### PR TITLE
Allow overbudget for small methods

### PR DESCRIPTION
**Pull request by EgorBo**
When we run out of time budget we start to give up even on simple properties, e.g. 
```csharp
    public readonly struct StandardFormat : IEquatable<StandardFormat>
    {
        private readonly byte _format;
        private readonly byte _precision;

        public char Symbol => (char)_format;
        public byte Precision => _precision;
        public bool HasPrecision => _precision != NoPrecision;
        ...
```
```
    [FAILED: inline exceeds budget] System.Buffers.StandardFormat:get_Symbol():ushort:this
    [FAILED: inline exceeds budget] System.Buffers.StandardFormat:get_PrecisionOrZero():ubyte:this
    [FAILED: inline exceeds budget] System.Buffers.StandardFormat:get_HasPrecision():bool:this
```

Presumably, for them we'll even save time if we inline them. In the worst case we might encounter a small method that has a call to another small method but that should be fine too + we have a hard limit on call graph depth.

@AndyAyersMS PTAL

Small diffs for BCL (more in real PGO runs), more hits in the 1P web service.

```
PMI CodeSize Diffs for System.Private.CoreLib.dll, framework assemblies [invoking .cctors] for  default jit

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 66224101
Total bytes of diff: 66223834
Total bytes of delta: -267 (-0.00 % of base)
Total relative delta: NaN
    diff is an improvement.
    relative diff is a regression.


Total byte diff includes 103 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    1 unique methods,      103 unique bytes

Top file regressions (bytes):
          79 : Microsoft.CodeAnalysis.dasm (0.00% of base)
          35 : Microsoft.CodeAnalysis.VisualBasic.dasm (0.00% of base)
          24 : Microsoft.CodeAnalysis.CSharp.dasm (0.00% of base)

Top file improvements (bytes):
        -306 : System.Private.CoreLib.dasm (-0.00% of base)
         -82 : System.Reflection.MetadataLoadContext.dasm (-0.04% of base)
         -11 : Microsoft.Extensions.DependencyInjection.dasm (-0.02% of base)
          -6 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm (-0.00% of base)

7 total files with Code Size differences (4 improved, 3 regressed), 269 unchanged.

Top method regressions (bytes):
         103 (     ∞ of base) : Microsoft.CodeAnalysis.dasm - Microsoft.CodeAnalysis.SmallDictionary`2[System.__Canon,int]:RightComplex(Microsoft.CodeAnalysis.SmallDictionary`2+AvlNode[System.__Canon,int]):Microsoft.CodeAnalysis.SmallDictionary`2+AvlNode[System.__Canon,int] (0 base, 1 diff methods)
          18 ( 5.63% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.ErrorFacts:CreateCategoriesMap():System.Collections.Immutable.ImmutableDictionary`2[int,System.String]
          18 ( 5.63% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.ErrorFactory:CreateCategoriesMap():System.Collections.Immutable.ImmutableDictionary`2[int,System.String]
          16 ( 6.93% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.CSharpExtensions:GetDirectives(Microsoft.CodeAnalysis.SyntaxNode,System.Func`2[Microsoft.CodeAnalysis.CSharp.Syntax.DirectiveTriviaSyntax,bool]):System.Collections.Generic.IList`1[Microsoft.CodeAnalysis.CSharp.Syntax.DirectiveTriviaSyntax]
          16 ( 6.93% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.VisualBasicExtensions:GetDirectives(Microsoft.CodeAnalysis.SyntaxNode,System.Func`2[Microsoft.CodeAnalysis.VisualBasic.Syntax.DirectiveTriviaSyntax,bool]):System.Collections.Generic.IList`1[Microsoft.CodeAnalysis.VisualBasic.Syntax.DirectiveTriviaSyntax]
           7 ( 2.89% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode:GetDirectives(System.Func`2[Microsoft.CodeAnalysis.CSharp.Syntax.DirectiveTriviaSyntax,bool]):System.Collections.Generic.IList`1[Microsoft.CodeAnalysis.CSharp.Syntax.DirectiveTriviaSyntax]:this
           7 ( 0.60% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Syntax.CompilationUnitSyntax:GetReferenceDirectives():System.Collections.Generic.IList`1[Microsoft.CodeAnalysis.CSharp.Syntax.ReferenceDirectiveTriviaSyntax]:this
           7 ( 2.89% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxNode:GetDirectives(System.Func`2[Microsoft.CodeAnalysis.VisualBasic.Syntax.DirectiveTriviaSyntax,bool]):System.Collections.Generic.IList`1[Microsoft.CodeAnalysis.VisualBasic.Syntax.DirectiveTriviaSyntax]:this

Top method improvements (bytes):
         -82 (-14.02% of base) : System.Reflection.MetadataLoadContext.dasm - System.Reflection.TypeLoading.Ecma.EcmaField:ComputeName():System.String:this
         -44 (-5.99% of base) : System.Private.CoreLib.dasm - System.Buffers.Text.Utf8Formatter:TryFormat(int,System.Span`1[ubyte],byref,System.Buffers.StandardFormat):bool
         -43 (-7.29% of base) : System.Private.CoreLib.dasm - System.Reflection.RuntimeModule:GetType(System.String,bool,bool):System.Type:this
         -41 (-6.75% of base) : System.Private.CoreLib.dasm - System.Buffers.Text.Utf8Formatter:TryFormat(ubyte,System.Span`1[ubyte],byref,System.Buffers.StandardFormat):bool
         -41 (-6.77% of base) : System.Private.CoreLib.dasm - System.Buffers.Text.Utf8Formatter:TryFormat(ushort,System.Span`1[ubyte],byref,System.Buffers.StandardFormat):bool
         -35 (-6.97% of base) : System.Private.CoreLib.dasm - System.Reflection.Emit.RuntimeAssemblyBuilder:GetType(System.String,bool,bool):System.Type:this
         -34 (-6.20% of base) : System.Private.CoreLib.dasm - System.Reflection.TypeNameParser:GetType(System.String,System.Reflection.Assembly,bool,bool):System.Type
         -34 (-5.73% of base) : System.Private.CoreLib.dasm - System.Type:GetType(System.String,bool,bool):System.Type
         -34 (-5.75% of base) : System.Private.CoreLib.dasm - System.Type:GetType(System.String,bool):System.Type
         -15 (-1.65% of base) : Microsoft.CodeAnalysis.dasm - Microsoft.CodeAnalysis.MetadataReaderExtensions:IsTheObjectClass(System.Reflection.Metadata.MetadataReader,System.Reflection.Metadata.TypeDefinition):bool
         -12 (-1.40% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Blender:ReadNode(int):Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.BlendedNode:this
         -12 (-1.40% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Blender:ReadToken(int):Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.BlendedNode:this
         -11 (-0.59% of base) : Microsoft.Extensions.DependencyInjection.dasm - Microsoft.Extensions.DependencyInjection.CallSiteJsonFormatter:VisitCallSiteMain(Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite,Microsoft.Extensions.DependencyInjection.CallSiteJsonFormatter+CallSiteFormatterContext):System.Object:this
          -9 (-0.98% of base) : Microsoft.CodeAnalysis.dasm - Microsoft.CodeAnalysis.Diagnostics.AnalysisResult:GetDiagnostics(System.Collections.Generic.IEnumerable`1[Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer]):System.Collections.Immutable.ImmutableArray`1[Microsoft.CodeAnalysis.Diagnostic]:this
          -6 (-0.37% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.Syntax.CompilationUnitSyntax:GetReferenceDirectives(System.Func`2[Microsoft.CodeAnalysis.VisualBasic.Syntax.ReferenceDirectiveTriviaSyntax,bool]):System.Collections.Generic.IList`1[Microsoft.CodeAnalysis.VisualBasic.Syntax.ReferenceDirectiveTriviaSyntax]:this
          -6 (-0.15% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.Parsers.Kernel.MemoryProcessMemInfoTraceData:ToXml(System.Text.StringBuilder):System.Text.StringBuilder:this

Top method regressions (percentages):
         103 (     ∞ of base) : Microsoft.CodeAnalysis.dasm - Microsoft.CodeAnalysis.SmallDictionary`2[System.__Canon,int]:RightComplex(Microsoft.CodeAnalysis.SmallDictionary`2+AvlNode[System.__Canon,int]):Microsoft.CodeAnalysis.SmallDictionary`2+AvlNode[System.__Canon,int] (0 base, 1 diff methods)
          16 ( 6.93% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.CSharpExtensions:GetDirectives(Microsoft.CodeAnalysis.SyntaxNode,System.Func`2[Microsoft.CodeAnalysis.CSharp.Syntax.DirectiveTriviaSyntax,bool]):System.Collections.Generic.IList`1[Microsoft.CodeAnalysis.CSharp.Syntax.DirectiveTriviaSyntax]
          16 ( 6.93% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.VisualBasicExtensions:GetDirectives(Microsoft.CodeAnalysis.SyntaxNode,System.Func`2[Microsoft.CodeAnalysis.VisualBasic.Syntax.DirectiveTriviaSyntax,bool]):System.Collections.Generic.IList`1[Microsoft.CodeAnalysis.VisualBasic.Syntax.DirectiveTriviaSyntax]
          18 ( 5.63% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.ErrorFacts:CreateCategoriesMap():System.Collections.Immutable.ImmutableDictionary`2[int,System.String]
          18 ( 5.63% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.ErrorFactory:CreateCategoriesMap():System.Collections.Immutable.ImmutableDictionary`2[int,System.String]
           7 ( 2.89% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode:GetDirectives(System.Func`2[Microsoft.CodeAnalysis.CSharp.Syntax.DirectiveTriviaSyntax,bool]):System.Collections.Generic.IList`1[Microsoft.CodeAnalysis.CSharp.Syntax.DirectiveTriviaSyntax]:this
           7 ( 2.89% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxNode:GetDirectives(System.Func`2[Microsoft.CodeAnalysis.VisualBasic.Syntax.DirectiveTriviaSyntax,bool]):System.Collections.Generic.IList`1[Microsoft.CodeAnalysis.VisualBasic.Syntax.DirectiveTriviaSyntax]:this
           7 ( 0.60% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Syntax.CompilationUnitSyntax:GetReferenceDirectives():System.Collections.Generic.IList`1[Microsoft.CodeAnalysis.CSharp.Syntax.ReferenceDirectiveTriviaSyntax]:this

Top method improvements (percentages):
         -82 (-14.02% of base) : System.Reflection.MetadataLoadContext.dasm - System.Reflection.TypeLoading.Ecma.EcmaField:ComputeName():System.String:this
         -43 (-7.29% of base) : System.Private.CoreLib.dasm - System.Reflection.RuntimeModule:GetType(System.String,bool,bool):System.Type:this
         -35 (-6.97% of base) : System.Private.CoreLib.dasm - System.Reflection.Emit.RuntimeAssemblyBuilder:GetType(System.String,bool,bool):System.Type:this
         -41 (-6.77% of base) : System.Private.CoreLib.dasm - System.Buffers.Text.Utf8Formatter:TryFormat(ushort,System.Span`1[ubyte],byref,System.Buffers.StandardFormat):bool
         -41 (-6.75% of base) : System.Private.CoreLib.dasm - System.Buffers.Text.Utf8Formatter:TryFormat(ubyte,System.Span`1[ubyte],byref,System.Buffers.StandardFormat):bool
         -34 (-6.20% of base) : System.Private.CoreLib.dasm - System.Reflection.TypeNameParser:GetType(System.String,System.Reflection.Assembly,bool,bool):System.Type
         -44 (-5.99% of base) : System.Private.CoreLib.dasm - System.Buffers.Text.Utf8Formatter:TryFormat(int,System.Span`1[ubyte],byref,System.Buffers.StandardFormat):bool
         -34 (-5.75% of base) : System.Private.CoreLib.dasm - System.Type:GetType(System.String,bool):System.Type
         -34 (-5.73% of base) : System.Private.CoreLib.dasm - System.Type:GetType(System.String,bool,bool):System.Type
         -15 (-1.65% of base) : Microsoft.CodeAnalysis.dasm - Microsoft.CodeAnalysis.MetadataReaderExtensions:IsTheObjectClass(System.Reflection.Metadata.MetadataReader,System.Reflection.Metadata.TypeDefinition):bool
         -12 (-1.40% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Blender:ReadNode(int):Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.BlendedNode:this
         -12 (-1.40% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Blender:ReadToken(int):Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.BlendedNode:this
          -9 (-0.98% of base) : Microsoft.CodeAnalysis.dasm - Microsoft.CodeAnalysis.Diagnostics.AnalysisResult:GetDiagnostics(System.Collections.Generic.IEnumerable`1[Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer]):System.Collections.Immutable.ImmutableArray`1[Microsoft.CodeAnalysis.Diagnostic]:this
         -11 (-0.59% of base) : Microsoft.Extensions.DependencyInjection.dasm - Microsoft.Extensions.DependencyInjection.CallSiteJsonFormatter:VisitCallSiteMain(Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite,Microsoft.Extensions.DependencyInjection.CallSiteJsonFormatter+CallSiteFormatterContext):System.Object:this
          -6 (-0.37% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.Syntax.CompilationUnitSyntax:GetReferenceDirectives(System.Func`2[Microsoft.CodeAnalysis.VisualBasic.Syntax.ReferenceDirectiveTriviaSyntax,bool]):System.Collections.Generic.IList`1[Microsoft.CodeAnalysis.VisualBasic.Syntax.ReferenceDirectiveTriviaSyntax]:this
          -6 (-0.15% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - Microsoft.Diagnostics.Tracing.Parsers.Kernel.MemoryProcessMemInfoTraceData:ToXml(System.Text.StringBuilder):System.Text.StringBuilder:this

24 total methods with Code Size differences (16 improved, 8 regressed), 414998 unchanged
```

(`103 (     ∞ of base) : Microsoft.CodeAnalysis.dasm` is a jit-diff random artifact)

### CODE REVIEW

File: src/coreclr/jit/inlinepolicy.cpp [hunk 1]
```diff
@@ -502,11 +502,25 @@ bool DefaultPolicy::BudgetCheck() const
         //
         assert(m_IsForceInlineKnown);
         assert(m_CallsiteDepth > 0);
-        const bool allowOverBudget = m_IsForceInline && (m_CallsiteDepth <= strategy->GetMaxForceInlineDepth());
+        bool allowOverBudget = m_IsForceInline && (m_CallsiteDepth <= strategy->GetMaxForceInlineDepth());
+
+        if (!allowOverBudget && (m_CodeSize <= (InlineStrategy::ALWAYS_INLINE_SIZE / 2)))
```
Comment (EgorBo) [inline src/coreclr/jit/inlinepolicy.cpp:507]: `ALWAYS_INLINE_SIZE / 2` (8 bytes) is probably too conservative, but I was targeting simple properties

File: src/coreclr/jit/inlinepolicy.cpp [hunk 2]
```diff
@@ -502,11 +502,25 @@ bool DefaultPolicy::BudgetCheck() const
...
+        {
+            // We're not going to inline no-return calls anyway
+            JITDUMP("Allowing over-budget for known no-returns\n")
+            allowOverBudget = true;
+        }
 
         if (allowOverBudget)
         {
-            JITDUMP("Allowing over-budget top-level forceinline\n");
+            JITDUMP("Allowing over-budget top-level forceinline\n")
```
Comment (AndyAyersMS) [inline src/coreclr/jit/inlinepolicy.cpp:523]:
```suggestion
            JITDUMP("Allowing over-budget: top-level forceinline, no return call, or small inlinee\n");
```

File: src/coreclr/jit/inlinepolicy.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/inlinepolicy.cpp:507]: Did you look at what happened for larger values?

**Comment by neon-sunset** [comment]
I don't have knowledge of how exactly inlining budgeting works but is it possible to exempt forceinline and small methods from budget calculation instead? E.g. large method body with many property accessors won't run out of budget no matter how big it is because inlining let's say below 16B is supposedly cheaper? In addition, is it possible to have even more relaxed restrictions for these for NativeAOT, similar to how it has increased inlining and time budget?

File: src/coreclr/jit/inlinepolicy.cpp (hunk 1 above)
Comment (EgorBo) [reply src/coreclr/jit/inlinepolicy.cpp:507]: 1 party web service stats:
Main: 2350 "exceed budget"
This PR: 2173 "exceed budget"
This PR + 16 bytes: 1865 "exceed budget"
This PR + 32 bytes: 1743 "exceed budget"
This PR + 64 bytes: 1743 "exceed budget" (same number)

Sounds like 16b is a good choice?

**Comment by EgorBo** [comment]
> I don't have knowledge of how exactly inlining budgeting works but is it possible to exempt forceinline and small methods from budget calculation instead?

We test various options now, should be very careful around ignoring the time budget

File: src/coreclr/jit/inlinepolicy.cpp (hunk 1 above)
Comment (EgorBo) [reply src/coreclr/jit/inlinepolicy.cpp]: Yep, it's 16

Comment (AndyAyersMS) [reply src/coreclr/jit/inlinepolicy.cpp]: Yes, sounds good.

**Review by AndyAyersMS — APPROVED** [review]

**Review by AndyAyersMS — APPROVED** [review]

**Pull request merged**